AWS/aws/event_aggregator.py

```diff
@@ -0,0 +1,169 @@
+        self.lock = threading.Lock()
+
+    def start_flush_on_ttl(self, ttl: int, on_flush_func: Callable[[dict], None], delay: float = 10):
+        """
+        This method MUST be called to trigger the execution of the ttl thread
+        """
+        self.ttl_thread = EventAggregatorTTLThread(
+            event_aggregator=self, ttl=ttl, on_flush_func=on_flush_func, delay=delay
+        )
+        self.ttl_thread.start()
+
+    def stop(self):
+        """
+        This method is only meant to stop the ttl thread
+        """
+        if self.ttl_thread:
+            self.ttl_thread.stop()
+            self.ttl_thread.join()
+
+    def flush_all(self) -> list[dict]:
+        """
+        returns all the aggregated events and remove them from the ongoing aggregations
+        """
+        aggregated_events = []
+
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for aggregation in self.aggregations.values():
+                if aggregation.count > 0:
+                    aggregated_events.append(aggregation.get_aggregated_event())
+
+            self.aggregations = dict()
+        return aggregated_events
+
+    def flush_all_ttl(self, ttl: int) -> list[dict]:
+        """
+        Returns (and delete) all the events we aggregate for at least the ttl time
+        """
+        event_hashes_to_flush = []
+        aggregated_events = []
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for event_hash, aggregation in self.aggregations.items():
+                if aggregation.start + datetime.timedelta(seconds=ttl) < datetime.datetime.utcnow():
+                    event_hashes_to_flush.append(event_hash)
+
+                    if aggregation.count > 0:
+                        aggregated_events.append(aggregation.get_aggregated_event())
+
+            for event_hash in event_hashes_to_flush:
+                del self.aggregations[event_hash]
+
+        return aggregated_events
+
+    def get_hash(self, event: dict) -> str | None:
+        """
+        Returns the hash to fingerprint the specified event
+        """
+        event_dialect_uuid = event["sekoiaio"]["intake"]["dialect_uuid"]
+
+        for fingerprint in self.aggregation_definitions.get(event_dialect_uuid, []):
+            if fingerprint.condition_func(event):
+                return f"{event_dialect_uuid};{fingerprint.build_hash_str_func(event)}"
+        return None
+
+    def aggregate(self, event: dict) -> dict | None:
+        try:
+            event_hash = self.get_hash(event)
+            # if no hash can be computed, we don't aggregate and forward the event
+            if not event_hash:
+                return event
+
+            # the aggregation stores the parsed timestamp
+            event_dt = parse_datetime_as_naive(event["@timestamp"])
+
+            # if hash is already known
+            with self.lock:
```

Comment:
So if I understand correctly, first occurrence is sent immediately and the aggregated one will follow once TTL is reached. Is that right ?

Reply:
yes exactly